File: mockserver-client-go/a2a.go
package mockserver

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"
)

// A2A (Agent2Agent) mock builder for the MockServer Go client.
//
// This is an idiomatic Go port of the Java client A2aMockBuilder
// (org.mockserver.client.A2aMockBuilder). It builds the set of MockServer
// expectations needed to mock an A2A agent: a GET agent-card discovery document
// plus the JSON-RPC 2.0 task methods (tasks/send, tasks/get, tasks/cancel) over
// HTTP, with optional SSE streaming and push-notification delivery. The produced
// expectation JSON is wire-compatible with the Java builder so both drive the
// same server behaviour.
//
// Usage:
//
//	mockserver.A2aMock("/a2a").
//	    WithAgentName("WeatherAgent").
//	    WithSkill("forecast", func(s *mockserver.A2aSkillBuilder) {
//	        s.WithName("Forecast").
//	            WithDescription("Provides weather forecasts").
//	            WithTag("weather").
//	            WithExample("What is the weather in London?")
//	    }).
//	    OnTaskSend("(?i).*london.*", "Sunny in London", false).
//	    ApplyTo(client)

// ---------------------------------------------------------------------------
// Internal definitions
// ---------------------------------------------------------------------------

type a2aSkill struct {
	id          string
	name        string
	hasName     bool
	description string
	hasDesc     bool
	tags        []string
	examples    []string
}

type a2aTaskHandler struct {
	messagePattern string
	responseText   string
	isError        bool
}

// ---------------------------------------------------------------------------
// A2aMockBuilder
// ---------------------------------------------------------------------------

// A2aMockBuilder builds the expectations that emulate an A2A agent.
type A2aMockBuilder struct {
	path                string
	agentCardPath       string
	agentName           string
	agentDescription    string
	agentVersion        string
	agentURL            string
	hasAgentURL         bool
	defaultTaskResponse string
	streaming           bool
	streamingMethod     string
	pushNotificationURL string
	hasPushNotification bool
	skills              []a2aSkill
	taskHandlers        []a2aTaskHandler
}

// A2aMock creates a new A2A mock builder. path defaults to "/a2a" when empty.
func A2aMock(path string) *A2aMockBuilder {
	if path == "" {
		path = "/a2a"
	}
	return &A2aMockBuilder{
		path:                path,
		agentCardPath:       "/.well-known/agent.json",
		agentName:           "MockAgent",
		agentDescription:    "A mock A2A agent",
		agentVersion:        "1.0.0",
		defaultTaskResponse: "Task completed successfully",
		streamingMethod:     "message/stream",
	}
}

// WithAgentName sets the agent name reported by the agent card.
func (b *A2aMockBuilder) WithAgentName(name string) *A2aMockBuilder {
	b.agentName = name
	return b
}

// WithAgentDescription sets the agent description reported by the agent card.
func (b *A2aMockBuilder) WithAgentDescription(description string) *A2aMockBuilder {
	b.agentDescription = description
	return b
}

// WithAgentVersion sets the agent version reported by the agent card.
func (b *A2aMockBuilder) WithAgentVersion(version string) *A2aMockBuilder {
	b.agentVersion = version
	return b
}

// WithAgentURL sets the agent URL reported by the agent card. When unset it
// defaults to "http://localhost" + path.
func (b *A2aMockBuilder) WithAgentURL(agentURL string) *A2aMockBuilder {
	b.agentURL = agentURL
	b.hasAgentURL = true
	return b
}

// WithAgentCardPath overrides the GET path of the agent-card discovery document
// (default "/.well-known/agent.json").
func (b *A2aMockBuilder) WithAgentCardPath(path string) *A2aMockBuilder {
	b.agentCardPath = path
	return b
}

// WithDefaultTaskResponse sets the text returned by tasks/send and tasks/get
// when no custom task handler matches.
func (b *A2aMockBuilder) WithDefaultTaskResponse(response string) *A2aMockBuilder {
	b.defaultTaskResponse = response
	return b
}

// WithStreaming advertises and mocks the A2A streaming capability. When enabled
// the agent card reports capabilities.streaming: true and the streaming JSON-RPC
// method (default "message/stream") returns an SSE stream of status/artifact
// update events wrapped in JSON-RPC 2.0 envelopes.
func (b *A2aMockBuilder) WithStreaming() *A2aMockBuilder {
	b.streaming = true
	return b
}

// WithStreamingMethod overrides the JSON-RPC method that triggers the streaming
// response (e.g. the legacy "tasks/sendSubscribe"). Implies WithStreaming.
func (b *A2aMockBuilder) WithStreamingMethod(method string) *A2aMockBuilder {
	b.streamingMethod = method
	b.streaming = true
	return b
}

// WithPushNotifications advertises and mocks A2A push notifications. When set the
// agent card reports capabilities.pushNotifications: true, the
// tasks/pushNotificationConfig/set method echoes the registered config, and each
// tasks/send additionally POSTs the completed task to the supplied webhook URL
// while still returning the JSON-RPC task response to the caller.
func (b *A2aMockBuilder) WithPushNotifications(webhookURL string) *A2aMockBuilder {
	b.pushNotificationURL = webhookURL
	b.hasPushNotification = true
	return b
}

// A2aSkillBuilder configures a single advertised A2A skill.
type A2aSkillBuilder struct {
	skill *a2aSkill
}

// WithName sets the skill name (defaults to the skill id when unset).
func (s *A2aSkillBuilder) WithName(name string) *A2aSkillBuilder {
	s.skill.name = name
	s.skill.hasName = true
	return s
}

// WithDescription sets the skill description.
func (s *A2aSkillBuilder) WithDescription(description string) *A2aSkillBuilder {
	s.skill.description = description
	s.skill.hasDesc = true
	return s
}

// WithTag appends a skill tag.
func (s *A2aSkillBuilder) WithTag(tag string) *A2aSkillBuilder {
	s.skill.tags = append(s.skill.tags, tag)
	return s
}

// WithExample appends a skill example prompt.
func (s *A2aSkillBuilder) WithExample(example string) *A2aSkillBuilder {
	s.skill.examples = append(s.skill.examples, example)
	return s
}

// WithSkill registers an advertised skill, configured via the supplied closure.
func (b *A2aMockBuilder) WithSkill(id string, configure func(*A2aSkillBuilder)) *A2aMockBuilder {
	sb := &A2aSkillBuilder{skill: &a2aSkill{id: id}}
	if configure != nil {
		configure(sb)
	}
	b.skills = append(b.skills, *sb.skill)
	return b
}

// OnTaskSend registers a custom tasks/send handler: when the first text part of
// the message matches messagePattern (a regular expression), the agent returns
// responseText with the given error flag. Handlers are matched (via JSONPath)
// before the default tasks/send response.
func (b *A2aMockBuilder) OnTaskSend(messagePattern, responseText string, isError bool) *A2aMockBuilder {
	if messagePattern == "" {
		messagePattern = ".*"
	}
	b.taskHandlers = append(b.taskHandlers, a2aTaskHandler{
		messagePattern: messagePattern,
		responseText:   responseText,
		isError:        isError,
	})
	return b
}

// ---------------------------------------------------------------------------
// Expectation assembly
// ---------------------------------------------------------------------------

func (b *A2aMockBuilder) buildAgentCard() Expectation {
	var skillsItems []string
	for _, skill := range b.skills {
		var sb strings.Builder
		sb.WriteString("{")
		sb.WriteString(`"id": "` + escapeJson(skill.id) + `"`)
		name := skill.id
		if skill.hasName {
			name = skill.name
		}
		sb.WriteString(`, "name": "` + escapeJson(name) + `"`)
		if skill.hasDesc {
			sb.WriteString(`, "description": "` + escapeJson(skill.description) + `"`)
		}
		if len(skill.tags) > 0 {
			tagItems := make([]string, 0, len(skill.tags))
			for _, tag := range skill.tags {
				tagItems = append(tagItems, `"`+escapeJson(tag)+`"`)
			}
			sb.WriteString(`, "tags": [` + strings.Join(tagItems, ", ") + "]")
		}
		if len(skill.examples) > 0 {
			exItems := make([]string, 0, len(skill.examples))
			for _, ex := range skill.examples {
				exItems = append(exItems, `"`+escapeJson(ex)+`"`)
			}
			sb.WriteString(`, "examples": [` + strings.Join(exItems, ", ") + "]")
		}
		sb.WriteString("}")
		skillsItems = append(skillsItems, sb.String())
	}
	skillsJson := "[" + strings.Join(skillsItems, ", ") + "]"

	agentURL := "http://localhost" + b.path
	if b.hasAgentURL {
		agentURL = b.agentURL
	}

	agentCardJson := "{" +
		`"name": "` + escapeJson(b.agentName) + `", ` +
		`"description": "` + escapeJson(b.agentDescription) + `", ` +
		`"version": "` + escapeJson(b.agentVersion) + `", ` +
		`"url": "` + escapeJson(agentURL) + `", ` +
		`"capabilities": {"streaming": ` + boolStr(b.streaming) +
		`, "pushNotifications": ` + boolStr(b.hasPushNotification) +
		`, "stateTransitionHistory": false}, ` +
		`"skills": ` + skillsJson + "}"

	return Expectation{
		HttpRequest: &HttpRequest{Method: "GET", Path: b.agentCardPath},
		HttpResponse: &HttpResponse{
			StatusCode: 200,
			Headers:    map[string][]string{"Content-Type": {"application/json"}},
			Body:       agentCardJson,
		},
	}
}

func (b *A2aMockBuilder) buildTasksSend() Expectation {
	resultJson := b.taskResultJson(b.defaultTaskResponse, false, true)
	return velocityTemplateExpectation(jsonRpcRequest(b.path, "tasks/send"), resultJson)
}

func (b *A2aMockBuilder) buildTasksGet() Expectation {
	resultJson := b.taskResultJson(b.defaultTaskResponse, false, true)
	return velocityTemplateExpectation(jsonRpcRequest(b.path, "tasks/get"), resultJson)
}

func (b *A2aMockBuilder) buildTasksCancel() Expectation {
	resultJson := `{"id": "mock-task-id", "status": {"state": "canceled"}}`
	return velocityTemplateExpectation(jsonRpcRequest(b.path, "tasks/cancel"), resultJson)
}

func (b *A2aMockBuilder) buildStreaming() Expectation {
	text := escapeJson(b.defaultTaskResponse)
	taskID := "mock-task-id"

	closeConnection := true
	sse := &HttpSseResponse{
		StatusCode: 200,
		Events: []SseEvent{
			{
				Event: "message",
				Data: `{"jsonrpc": "2.0", "id": "1", "result": ` +
					`{"taskId": "` + taskID + `", "kind": "status-update", ` +
					`"status": {"state": "working"}, "final": false}}`,
			},
			{
				Event: "message",
				Data: `{"jsonrpc": "2.0", "id": "1", "result": ` +
					`{"taskId": "` + taskID + `", "kind": "artifact-update", ` +
					`"artifact": {"parts": [{"type": "text", "text": "` + text + `"}]}}}`,
			},
			{
				Event: "message",
				Data: `{"jsonrpc": "2.0", "id": "1", "result": ` +
					`{"taskId": "` + taskID + `", "kind": "status-update", ` +
					`"status": {"state": "completed"}, "final": true}}`,
			},
		},
		CloseConnection: &closeConnection,
	}

	return Expectation{
		HttpRequest:     jsonRpcRequest(b.path, b.streamingMethod),
		HttpSseResponse: sse,
	}
}

func (b *A2aMockBuilder) buildPushNotificationConfig() Expectation {
	// Echo the registered push-notification config back as the JSON-RPC result.
	resultJson := `{"url": "` + escapeVelocity(escapeJson(b.pushNotificationURL)) + `"}`
	return velocityTemplateExpectation(jsonRpcRequest(b.path, "tasks/pushNotificationConfig/set"), resultJson)
}

func (b *A2aMockBuilder) buildPushNotificationDelivery() Expectation {
	// A tasks/send both returns the JSON-RPC task response to the caller AND POSTs
	// the completed task to the configured webhook. Modelled as an
	// override-forwarded-request: the request override targets the webhook (literal
	// body, JSON-escaped only), and a Velocity response template produces the
	// caller's JSON-RPC response so the request's id is echoed back.
	target := parseWebhookTarget(b.pushNotificationURL)

	pushBody := `{"jsonrpc": "2.0", "result": ` + b.taskResultJson(b.defaultTaskResponse, false, false) + "}"

	secure := target.secure
	scheme := "HTTP"
	if secure {
		scheme = "HTTPS"
	}
	webhookRequest := &HttpRequest{
		Method:        "POST",
		Path:          target.path,
		SocketAddress: &SocketAddress{Host: target.host, Port: target.port, Scheme: scheme},
		Secure:        &secure,
		Headers: map[string][]string{
			"Host":         {target.hostHeader()},
			"Content-Type": {"application/json"},
		},
		Body: pushBody,
	}

	clientResponseTemplate := &HttpTemplate{
		TemplateType: "VELOCITY",
		Template:     velocityJsonRpcResponse(b.taskResultJson(b.defaultTaskResponse, false, true)),
	}

	return Expectation{
		HttpRequest: jsonRpcRequest(b.path, "tasks/send"),
		HttpOverrideForwardedRequest: &HttpOverrideForwardedRequest{
			RequestOverride:  webhookRequest,
			ResponseTemplate: clientResponseTemplate,
		},
	}
}

func (b *A2aMockBuilder) buildCustomTaskHandler(handler a2aTaskHandler) Expectation {
	escapedPattern := escapeMessagePattern(handler.messagePattern)
	jsonPath := "$[?(@.method == 'tasks/send' && @.params.message.parts[0].text =~ /" + escapedPattern + "/)]"
	resultJson := b.taskResultJson(handler.responseText, handler.isError, true)
	return velocityTemplateExpectation(jsonPathRequest(b.path, jsonPath), resultJson)
}

// escapeMessagePattern neutralises the characters that could let a
// user-supplied messagePattern (a regular expression) break out of the
// /.../ regex literal embedded in the JSONPath matcher, while preserving the
// pattern's own regex escape sequences (e.g. \d, \/, \\).
//
// The scan is single-pass and left-to-right over BYTES — every special
// character handled here is ASCII, and any other byte (including the
// continuation bytes of multi-byte UTF-8 runes) is emitted verbatim, so valid
// UTF-8 input is preserved unchanged.
//
//   - On a backslash: if another byte follows, the backslash and that byte are
//     both emitted verbatim so an existing escape sequence (\d, \/, \\, ...) is
//     left intact; a TRAILING backslash with nothing to escape is doubled to
//     \\ so it cannot escape the closing "/" delimiter and break out.
//   - A bare "/" is escaped to \/ so it cannot close the regex literal early.
//   - Newline / carriage-return are converted to \n / \r.
//   - A NUL byte is stripped.
//   - Any other byte is emitted verbatim.
//
// For every input that contains none of these breakout cases the output is
// byte-identical to the previous inline escaping.
func escapeMessagePattern(pattern string) string {
	var b strings.Builder
	for i := 0; i < len(pattern); i++ {
		c := pattern[i]
		switch {
		case c == '\\':
			if i+1 < len(pattern) {
				b.WriteByte('\\')
				b.WriteByte(pattern[i+1])
				i++
			} else {
				b.WriteString(`\\`)
			}
		case c == '/':
			b.WriteString(`\/`)
		case c == '\n':
			b.WriteString(`\n`)
		case c == '\r':
			b.WriteString(`\r`)
		case c == 0x00:
			// strip NUL
		default:
			b.WriteByte(c)
		}
	}
	return b.String()
}

// taskResultJson builds the A2A task result object. When velocity is true the
// text is escaped for survival through the Velocity engine (templated response
// bodies); when false only JSON escaping is applied (literal webhook bodies).
func (b *A2aMockBuilder) taskResultJson(responseText string, isError, velocity bool) string {
	escaped := escapeJson(responseText)
	if velocity {
		escaped = escapeVelocity(escaped)
	}
	state := "completed"
	if isError {
		state = "failed"
	}
	return `{"id": "mock-task-id", ` +
		`"status": {"state": "` + state + `"}, ` +
		`"artifacts": [{"parts": [{"type": "text", "text": "` + escaped + `"}]}]}`
}

// Build returns the ordered list of expectations that emulate the A2A agent.
func (b *A2aMockBuilder) Build() []Expectation {
	expectations := []Expectation{b.buildAgentCard()}

	for _, handler := range b.taskHandlers {
		expectations = append(expectations, b.buildCustomTaskHandler(handler))
	}

	if b.streaming {
		expectations = append(expectations, b.buildStreaming())
	}

	if b.hasPushNotification {
		expectations = append(expectations,
			b.buildPushNotificationConfig(),
			b.buildPushNotificationDelivery())
	} else {
		expectations = append(expectations, b.buildTasksSend())
	}
	expectations = append(expectations, b.buildTasksGet(), b.buildTasksCancel())

	return expectations
}

// ApplyTo registers the A2A expectations with the client via Upsert.
func (b *A2aMockBuilder) ApplyTo(c *Client) ([]Expectation, error) {
	return c.Upsert(b.Build()...)
}

// ---------------------------------------------------------------------------
// Webhook target parsing (mirrors the Java WebhookTarget)
// ---------------------------------------------------------------------------

type webhookTarget struct {
	host   string
	port   int
	secure bool
	path   string
}

func (t webhookTarget) hostHeader() string {
	return t.host + ":" + strconv.Itoa(t.port)
}

// parseWebhookTarget parses a push-notification webhook URL into its components.
// It panics on an invalid URL (no host), mirroring the throwing Java builder.
func parseWebhookTarget(raw string) webhookTarget {
	u, err := url.Parse(raw)
	if err != nil || u.Host == "" {
		panic(fmt.Sprintf("mockserver: invalid push-notification webhook URL: %q", raw))
	}
	secure := strings.EqualFold(u.Scheme, "https")
	host := u.Hostname()
	if host == "" {
		panic(fmt.Sprintf("mockserver: invalid push-notification webhook URL (no host): %q", raw))
	}
	port := 80
	if secure {
		port = 443
	}
	if p := u.Port(); p != "" {
		if parsed, perr := strconv.Atoi(p); perr == nil {
			port = parsed
		}
	}
	path := u.EscapedPath()
	if path == "" {
		path = "/"
	}
	return webhookTarget{host: host, port: port, secure: secure, path: path}
}

func boolStr(v bool) string {
	if v {
		return "true"
	}
	return "false"
}

File: mockserver-client-ruby/lib/mockserver/a2a.rb
# frozen_string_literal: true

require 'json'
require 'uri'

module MockServer
  # Fluent builder for mocking an A2A (Agent-to-Agent) agent.
  #
  # Mirrors the Java +A2aMockBuilder+ (and its Node/Python siblings). It produces
  # the same wire-level expectation JSON: a set of HTTP expectations that emulate
  # an A2A agent serving an agent card and speaking JSON-RPC 2.0 over +POST
  # <path>+. The generated control-plane expectations are:
  #
  #   * agent-card +GET <agent_card_path>+ (default +/.well-known/agent.json+);
  #   * JSON-RPC +tasks/send+, +tasks/get+, +tasks/cancel+;
  #   * optional SSE streaming (+with_streaming+);
  #   * optional push-notification config + delivery (+with_push_notifications+);
  #   * optional per-message custom task handlers (+on_task_send+);
  #   * optional advertised skills (+with_skill+).
  #
  # Each JSON-RPC response is a Velocity template that echoes the incoming
  # JSON-RPC id back via +$!{request.jsonRpcRawId}+.
  #
  # @example
  #   MockServer::A2A.a2a_mock('/a2a')
  #     .with_agent_name('WeatherAgent')
  #     .with_skill('weather')
  #       .with_name('Weather lookup')
  #       .with_tag('forecast')
  #     .and_then
  #     .on_task_send
  #       .matching_message('forecast')
  #       .responding_with('Sunny, 25C')
  #     .and_then
  #     .apply_to(client)
  module A2A
    # JSON-escape a string the same way Jackson +writeValueAsString+ does, then
    # strip the surrounding quotes — yielding only the escaped inner content.
    # @api private
    def self.escape_json(value)
      return '' if value.nil?

      quoted = JSON.generate(value.to_s)
      quoted[1...-1]
    end

    # Escape Velocity metacharacters so literal +$+ / +#+ survive rendering.
    # @api private
    def self.escape_velocity(value)
      return value if value.nil?

      value.to_s.gsub('$', '${esc.d}').gsub('#', '${esc.h}')
    end

    # @api private
    def self.velocity_json_rpc_response(result_json)
      '{"statusCode": 200, ' \
        '"headers": [{"name": "Content-Type", "values": ["application/json"]}], ' \
        '"body": {"jsonrpc": "2.0", "result": ' + result_json + ', "id": $!{request.jsonRpcRawId}}}'
    end

    # @api private
    # Wraps a Hash so it responds to +to_h+, allowing it to be passed to
    # +Client#upsert+.
    class RawExpectation
      def initialize(hash)
        @hash = hash
      end

      def to_h
        @hash
      end
    end

    # @api private
    SkillDef = Struct.new(:id, :name, :description, :tags, :examples)
    # @api private
    TaskHandler = Struct.new(:message_pattern, :response_text, :is_error)

    # Resolved components of a push-notification webhook URL.
    # @api private
    WebhookTarget = Struct.new(:host, :port, :secure, :path) do
      def host_header
        "#{host}:#{port}"
      end

      def self.parse(url)
        uri = URI.parse(url)
        secure = uri.scheme.to_s.downcase == 'https'
        host = uri.host
        raise ArgumentError, "Invalid push-notification webhook URL (no host): #{url}" if host.nil? || host.empty?

        port = uri.port || (secure ? 443 : 80)
        path = uri.path
        path = '/' if path.nil? || path.empty?
        new(host, port, secure, path)
      end
    end

    class A2aMockBuilder
      def initialize(path = '/a2a')
        @path = path.is_a?(String) ? path : '/a2a'
        @agent_card_path = '/.well-known/agent.json'
        @agent_name = 'MockAgent'
        @agent_description = 'A mock A2A agent'
        @agent_version = '1.0.0'
        @agent_url = nil
        @skills = []
        @task_handlers = []
        @default_task_response = 'Task completed successfully'
        @streaming = false
        @streaming_method = 'message/stream'
        @push_notification_url = nil
      end

      # @return [self]
      def with_agent_name(name)
        @agent_name = name
        self
      end

      # @return [self]
      def with_agent_description(description)
        @agent_description = description
        self
      end

      # @return [self]
      def with_agent_version(version)
        @agent_version = version
        self
      end

      # @return [self]
      def with_agent_url(url)
        @agent_url = url
        self
      end

      # @return [self]
      def with_agent_card_path(path)
        @agent_card_path = path
        self
      end

      # @return [self]
      def with_default_task_response(response)
        @default_task_response = response
        self
      end

      # Advertise + mock the A2A streaming capability. The agent card reports
      # +capabilities.streaming: true+ and the streaming JSON-RPC method (default
      # +message/stream+) returns an SSE stream of status/artifact update events.
      # @return [self]
      def with_streaming
        @streaming = true
        self
      end

      # Override the JSON-RPC method that triggers the streaming response.
      # Implies {#with_streaming}.
      # @return [self]
      def with_streaming_method(method)
        @streaming_method = method
        @streaming = true
        self
      end

      # Advertise + mock A2A push notifications. The agent card reports
      # +capabilities.pushNotifications: true+, +tasks/pushNotificationConfig/set+
      # echoes the registered config, and each +tasks/send+ additionally POSTs the
      # completed task to +webhook_url+ while still returning the JSON-RPC task
      # response to the caller.
      # @return [self]
      def with_push_notifications(webhook_url)
        @push_notification_url = webhook_url
        self
      end

      # @return [A2aSkillBuilder]
      def with_skill(id)
        A2aSkillBuilder.new(self, id)
      end

      # @return [A2aTaskHandlerBuilder]
      def on_task_send
        A2aTaskHandlerBuilder.new(self)
      end

      # @api private
      def add_skill(skill)
        @skills << skill
      end

      # @api private
      def add_task_handler(handler)
        @task_handlers << handler
      end

      # @return [Array<Hash>] the ordered list of expectations
      def build
        expectations = [build_agent_card_expectation]

        @task_handlers.each { |handler| expectations << build_custom_task_handler(handler) }

        expectations << build_streaming_expectation if @streaming

        if @push_notification_url
          expectations << build_push_notification_config_expectation
          expectations << build_push_notification_delivery_expectation
        else
          expectations << build_tasks_send_expectation
        end
        expectations << build_tasks_get_expectation
        expectations << build_tasks_cancel_expectation

        expectations
      end

      # @return [Array<Expectation>]
      def apply_to(client)
        client.upsert(*build.map { |h| RawExpectation.new(h) })
      end

      private

      def json_rpc_request(method)
        { 'method' => 'POST', 'path' => @path, 'body' => { 'type' => 'JSON_RPC', 'method' => method } }
      end

      def json_path_request(json_path)
        { 'method' => 'POST', 'path' => @path, 'body' => { 'type' => 'JSON_PATH', 'jsonPath' => json_path } }
      end

      def velocity_template_response_expectation(http_request, result_json)
        {
          'httpRequest' => http_request,
          'httpResponseTemplate' => {
            'template' => A2A.velocity_json_rpc_response(result_json),
            'templateType' => 'VELOCITY'
          }
        }
      end

      def build_agent_card_expectation
        skill_items = @skills.map do |skill|
          parts = ['{"id": "' + esc_plain(skill.id) + '"']
          parts << ', "name": "' + esc_plain(skill.name || skill.id) + '"'
          parts << ', "description": "' + esc_plain(skill.description) + '"' unless skill.description.nil?
          unless skill.tags.empty?
            parts << ', "tags": [' + skill.tags.map { |t| '"' + esc_plain(t) + '"' }.join(', ') + ']'
          end
          unless skill.examples.empty?
            parts << ', "examples": [' + skill.examples.map { |e| '"' + esc_plain(e) + '"' }.join(', ') + ']'
          end
          parts << '}'
          parts.join
        end
        skills_json = '[' + skill_items.join(', ') + ']'

        url = @agent_url || ('http://localhost' + @path)

        agent_card_json =
          '{' \
          '"name": "' + esc_plain(@agent_name) + '", ' \
          '"description": "' + esc_plain(@agent_description) + '", ' \
          '"version": "' + esc_plain(@agent_version) + '", ' \
          '"url": "' + esc_plain(url) + '", ' \
          '"capabilities": {"streaming": ' + bool(@streaming) + ', ' \
          '"pushNotifications": ' + bool(!@push_notification_url.nil?) + ', ' \
          '"stateTransitionHistory": false}, ' \
          '"skills": ' + skills_json + '}'

        {
          'httpRequest' => { 'method' => 'GET', 'path' => @agent_card_path },
          'httpResponse' => {
            'statusCode' => 200,
            'headers' => [{ 'name' => 'Content-Type', 'values' => ['application/json'] }],
            'body' => agent_card_json
          }
        }
      end

      def build_tasks_send_expectation
        velocity_template_response_expectation(
          json_rpc_request('tasks/send'),
          task_result_json(@default_task_response, false)
        )
      end

      def build_tasks_get_expectation
        velocity_template_response_expectation(
          json_rpc_request('tasks/get'),
          task_result_json(@default_task_response, false)
        )
      end

      def build_tasks_cancel_expectation
        result_json = '{"id": "mock-task-id", "status": {"state": "canceled"}}'
        velocity_template_response_expectation(json_rpc_request('tasks/cancel'), result_json)
      end

      def build_custom_task_handler(handler)
        escaped_pattern = escape_message_pattern(handler.message_pattern)
        json_path = "$[?(@.method == 'tasks/send' && @.params.message.parts[0].text =~ /" + escaped_pattern + '/)]'
        velocity_template_response_expectation(
          json_path_request(json_path),
          task_result_json(handler.response_text, handler.is_error)
        )
      end

      # Neutralize only the regex-delimiter breakout while preserving every
      # existing regex escape sequence. The user-supplied message_pattern is
      # documented as a regular expression and is embedded between `/.../`
      # delimiters inside a JSONPath. A single-pass scanner is required (rather
      # than independent gsubs) so that a backslash and the character it escapes
      # are consumed together — otherwise a trailing lone backslash, or an input
      # containing `\/`, could escape the closing `/` delimiter and break out of
      # the regex literal into the surrounding JSONPath/JSON.
      # (CodeQL rb/incomplete-sanitization, alert #65.)
      def escape_message_pattern(pattern)
        chars = pattern.to_s.chars
        out = +''
        i = 0
        while i < chars.length
          c = chars[i]
          case c
          when '\\'
            if i + 1 < chars.length
              out << '\\' << chars[i + 1] # preserve escape sequence (\d, \/, \\)
              i += 1                       # extra advance past the escaped char
            else
              out << '\\\\'                # trailing lone backslash -> literal backslash
            end
          when '/'  then out << '\\/'
          when "\n" then out << '\\n'
          when "\r" then out << '\\r'
          when "\0" then nil               # strip NUL
          else out << c
          end
          i += 1
        end
        out
      end

      def build_streaming_expectation
        text = A2A.escape_json(@default_task_response)
        task_id = 'mock-task-id'

        events = [
          { 'event' => 'message',
            'data' => '{"jsonrpc": "2.0", "id": "1", "result": ' \
              '{"taskId": "' + task_id + '", "kind": "status-update", ' \
              '"status": {"state": "working"}, "final": false}}' },
          { 'event' => 'message',
            'data' => '{"jsonrpc": "2.0", "id": "1", "result": ' \
              '{"taskId": "' + task_id + '", "kind": "artifact-update", ' \
              '"artifact": {"parts": [{"type": "text", "text": "' + text + '"}]}}}' },
          { 'event' => 'message',
            'data' => '{"jsonrpc": "2.0", "id": "1", "result": ' \
              '{"taskId": "' + task_id + '", "kind": "status-update", ' \
              '"status": {"state": "completed"}, "final": true}}' }
        ]

        {
          'httpRequest' => json_rpc_request(@streaming_method),
          'httpSseResponse' => {
            'statusCode' => 200,
            'events' => events,
            'closeConnection' => true
          }
        }
      end

      def build_push_notification_config_expectation
        # Echo the registered push-notification config back as the JSON-RPC result.
        result_json = '{"url": "' + A2A.escape_velocity(A2A.escape_json(@push_notification_url)) + '"}'
        velocity_template_response_expectation(
          json_rpc_request('tasks/pushNotificationConfig/set'),
          result_json
        )
      end

      def build_push_notification_delivery_expectation
        # A tasks/send both returns the JSON-RPC task response to the caller AND
        # POSTs the completed task to the configured webhook. Modelled with an
        # override-forwarded-request: the request override targets the webhook
        # (literal body, JSON-escaped only — no Velocity engine runs over a request
        # override), and a Velocity response *template* produces the caller's
        # JSON-RPC response so the request's id is echoed back.
        target = WebhookTarget.parse(@push_notification_url)

        push_body = '{"jsonrpc": "2.0", "result": ' + task_result_json_raw(@default_task_response, false) + '}'

        webhook_request = {
          'method' => 'POST',
          'path' => target.path,
          'socketAddress' => {
            'host' => target.host,
            'port' => target.port,
            'scheme' => target.secure ? 'HTTPS' : 'HTTP'
          },
          'secure' => target.secure,
          'headers' => [
            { 'name' => 'Host', 'values' => [target.host_header] },
            { 'name' => 'Content-Type', 'values' => ['application/json'] }
          ],
          'body' => push_body
        }

        {
          'httpRequest' => json_rpc_request('tasks/send'),
          'httpOverrideForwardedRequest' => {
            'requestOverride' => webhook_request,
            'responseTemplate' => {
              'template' => A2A.velocity_json_rpc_response(task_result_json(@default_task_response, false)),
              'templateType' => 'VELOCITY'
            }
          }
        }
      end

      # Velocity-templated result body: text must survive the Velocity engine, so
      # both JSON and Velocity escaping are applied (Velocity un-escapes at render).
      def task_result_json(response_text, is_error)
        task_result_json_with_text(A2A.escape_velocity(A2A.escape_json(response_text)), is_error)
      end

      # Literal (non-templated) result body (e.g. the webhook payload): only JSON
      # escaping — Velocity escaping would corrupt '$' / '#'.
      def task_result_json_raw(response_text, is_error)
        task_result_json_with_text(A2A.escape_json(response_text), is_error)
      end

      def task_result_json_with_text(escaped_text, is_error)
        state = is_error ? 'failed' : 'completed'
        '{"id": "mock-task-id", ' \
        '"status": {"state": "' + state + '"}, ' \
        '"artifacts": [{"parts": [{"type": "text", "text": "' + escaped_text + '"}]}]}'
      end

      def bool(value)
        value ? 'true' : 'false'
      end

      # JSON-escape then Velocity-escape, matching the reference clients (used for
      # the agent card, which is a literal — but kept Velocity-safe for parity with
      # the Java builder's escapeJson-only card; here the card is a literal HTTP
      # response body, so only JSON escaping is required).
      def esc_plain(value)
        A2A.escape_json(value)
      end
    end

    class A2aSkillBuilder
      def initialize(parent, id)
        @parent = parent
        @skill = SkillDef.new(id, nil, nil, [], [])
      end

      # @return [self]
      def with_name(name)
        @skill.name = name
        self
      end

      # @return [self]
      def with_description(description)
        @skill.description = description
        self
      end

      # @return [self]
      def with_tag(tag)
        @skill.tags << tag
        self
      end

      # @return [self]
      def with_example(example)
        @skill.examples << example
        self
      end

      # Commit the skill and return to the root builder.
      # @return [A2aMockBuilder]
      def and_then
        @parent.add_skill(@skill)
        @parent
      end
      alias and_ and_then
    end

    class A2aTaskHandlerBuilder
      def initialize(parent)
        @parent = parent
        @handler = TaskHandler.new('.*', 'Task completed', false)
      end

      # @return [self]
      def matching_message(pattern)
        @handler.message_pattern = pattern
        self
      end

      # @return [self]
      def responding_with(text, is_error = false)
        @handler.response_text = text
        @handler.is_error = is_error
        self
      end

      # Commit the handler and return to the root builder.
      # @return [A2aMockBuilder]
      def and_then
        @parent.add_task_handler(@handler)
        @parent
      end
      alias and_ and_then
    end

    # Create a new A2A mock builder. +path+ defaults to +/a2a+.
    # @return [A2aMockBuilder]
    def self.a2a_mock(path = '/a2a')
      A2aMockBuilder.new(path)
    end
  end
end
